### PR TITLE
*: Support delete in new plan

### PR DESCRIPTION
**Pull request by shenli**
@coocood @qiuyesuifeng

### CODE REVIEW

File: executor/converter/convert_stmt.go [hunk 1]
```diff
@@ -112,9 +112,11 @@ func convertDelete(converter *expressionConverter, v *ast.DeleteStmt) (*stmts.De
 		return nil, errors.Trace(err)
 	}
 	oldDelete.Refs = oldRefs
-	for _, val := range v.Tables {
-		tableIdent := table.Ident{Schema: val.Schema, Name: val.Name}
-		oldDelete.TableIdents = append(oldDelete.TableIdents, tableIdent)
+	if v.Tables != nil {
```
Comment (qiuyesuifeng) [inline executor/converter/convert_stmt.go]: Seems v.Tables should always be not nil in parser.y?

File: optimizer/plan/plans.go [hunk 2]
```diff
@@ -461,6 +461,30 @@ func (p *Update) Accept(v Visitor) (Plan, bool) {
 	return v.Leave(p)
 }
 
+// Delete represents a delete plan.
+type Delete struct {
+	basePlan
+
+	SelectPlan Plan
+	Tables     []*ast.TableName
+	MultiTable bool
```
Comment (ngaut) [inline optimizer/plan/plans.go]: s/MultiTable/IsMultiTable  ?

File: executor/converter/convert_stmt.go (hunk 1 above)
Comment (shenli) [inline executor/converter/convert_stmt.go]: For single-table mode, it is nil. See: https://github.com/pingcap/tidb/blob/master/parser/parser.y#L1199

File: executor/executor_write.go [hunk 3]
```diff
@@ -234,3 +236,101 @@ func (e *UpdateExec) Fields() []*ast.ResultField {
...
+// See: https://dev.mysql.com/doc/refman/5.7/en/delete.html
+type DeleteExec struct {
+	SelectExec Executor
+
+	ctx        context.Context
+	Tables     []*ast.TableName
+	MultiTable bool
+}
+
+// Next implements Execution Next interface.
```
Comment (coocood) [inline executor/executor_write.go]: Executor

File: optimizer/resolver.go [hunk 4]
```diff
@@ -243,9 +249,23 @@ func (nr *nameResolver) handleTableName(tn *ast.TableName) {
 	if tn.Schema.L == "" {
 		tn.Schema = nr.DefaultSchema
 	}
+	ctx := nr.currentContext()
+	if ctx != nil && ctx.inDeleteTableList {
```
Comment (coocood) [inline optimizer/resolver.go]: When will `ctx` be nil?

File: optimizer/resolver.go [hunk 5]
```diff
@@ -243,9 +249,23 @@ func (nr *nameResolver) handleTableName(tn *ast.TableName) {
...
+		idx, ok := ctx.tableMap[tn.Name.L]
```
Comment (coocood) [inline optimizer/resolver.go]: Use `tableUniqueName(tn.Schema, tn.Name))` instead of `tn.Name.L`.

File: executor/executor_write.go [hunk 6]
```diff
@@ -234,3 +236,101 @@ func (e *UpdateExec) Fields() []*ast.ResultField {
...
+func (e *DeleteExec) Next() (*Row, error) {
```
Comment (coocood) [inline executor/executor_write.go]: The `Next` method should delete one row per call as `UpdateExec` does.

File: optimizer/resolver.go (hunk 4 above)
Comment (shenli) [inline optimizer/resolver.go]: I am not sure. But some test cases meet panic because of this.

File: executor/executor_write.go (hunk 6 above)
Comment (shenli) [inline executor/executor_write.go]: It is not necessary. We can finish the job in one call.

Comment (coocood) [inline executor/executor_write.go]: It's not necessary, but it behave the same as the method name indicates `Next`.

Comment (coocood) [inline executor/executor_write.go]: At least, add a `finished` flag to prevent it from executing multiple times if `Next` is called multiple times.

**Comment by shenli** [comment]
PTAL @coocood @qiuyesuifeng

**Comment by coocood** [comment]
LGTM

File: optimizer/plan/planbuilder.go [hunk 7]
```diff
@@ -647,6 +649,32 @@ func (b *planBuilder) buildUpdateLists(list []*ast.Assignment, fields []*ast.Res
...
+			return nil
+		}
+	}
+	var tables []*ast.TableName
+	if del.Tables != nil {
+		tables = del.Tables.Tables
+	}
+	return &Delete{
+		Tables:       tables,
+		IsMultiTable: del.MultiTable,
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder.go]: Should we rename DeleteStmt's MultiTable to IsMultiTable to unify them?

Comment (ngaut) [inline optimizer/plan/planbuilder.go]: +1

Comment (shenli) [inline optimizer/plan/planbuilder.go]: ok

**Comment by shenli** [comment]
@qiuyesuifeng  PTAL

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**